Dockerization for 03-rwslotmachine

cosmintanasa47 opened the pull request:
I think it works. Tested the containers and worked on my computer. Disk usage seems to be too high. May need improvements. Don't know if this is enough to do the challenge/activity.

# Prerequisite Checklist

<!--
Please mark items appropriately:
-->

- [x] Read the [contribution guidelines](https://github.com/open-education-hub/binary-security/blob/main/CONTRIBUTING.md#pull-requests) regarding submitting new changes to the project;
- [x] Tested your changes against relevant architectures and platforms;
- [x] Updated relevant documentation (if needed).

## Description of changes

Added a Dockerfile that creates the environment to build and run the executable.
Added a Makefile to handle the Dockerfile (build, run, etc.)
Added a Makefile that is uploaded to the container and builds the executable, checks for the existence of the executable and checks if the executable is operating on port 31346.

With rwslotmachine3_runtime-cnt container running, I tested the communication with the app using nc and this was the result:

nc 127.0.0.1 32000
==== www / rww ====
1. Read (1/1)
2. Write (1/1)
3. Exit
|>2
Input slot index: 3
Input new slot value: 3
Slot[3]:= 0003
==== www / rww ====
1. Read (unavailable)
2. Write (unavailable)
3. Exit
|> 1
==== www / rww ====
1. Read (unavailable)
2. Write (unavailable)
3. Exit
|> 3

nc 127.0.0.1 32000
==== www / rww ====
1. Read (1/1)
2. Write (1/1)
3. Exit
|> 1
Input slot index: 3
Slot[3]: 0000
==== www / rww ====
1. Read (unavailable)
2. Write (unavailable)
3. Exit
|> 3

32000 - local port linked with container port 31346.

<!--
Please provide a detailed description of the changes made in this new PR.
-->

cosmintanasa47 committed:
03-rwslotmachine: Add Dockerfile and Makefile for 03-rwslotmachine dockerization

Add a multi-stage Dockerfile that builds the 32bit binary for
the rwslotmachine3.c file and runs the binary on port 31346.
Local port for build container is 32000 and the
runtime container uses port 32001.

Add a Makefile that handles the Dockerfile's build, run and stop.

It works on my computer. Disk usage seems to be too high.
May need improvements.

Resolves: #50

Signed-off-by: Cosmin Tanasa <tanasa.cosminandrei@gmail.com>

razvand commented on chapters/mitigations-and-defensive-strategies/defense-mechanisms/activities/03-rwslotmachine3/sol/Makefile:
```diff
@@ -0,0 +1,22 @@
+CFLAGS = -m32 -fno-stack-protector -Wall -Wno-format-security -no-pie
+
+.PHONY: build check_binary check_port
+
+
```
```suggestion

```

Remove duplicate blank line.

razvand commented on chapters/mitigations-and-defensive-strategies/defense-mechanisms/activities/03-rwslotmachine3/sol/Makefile:
```diff
@@ -0,0 +1,22 @@
+
+
+build:
+
```
```suggestion
```

razvand commented on chapters/mitigations-and-defensive-strategies/defense-mechanisms/activities/03-rwslotmachine3/sol/Makefile:
```diff
@@ -0,0 +1,22 @@
+
+build:
+
+	gcc $(CFLAGS) rwslotmachine3.c -o rwslotmachine3
```
```suggestion
	$(CC) $(CFLAGS) rwslotmachine3.c -o rwslotmachine3
```

razvand commented on chapters/mitigations-and-defensive-strategies/defense-mechanisms/activities/03-rwslotmachine3/sol/Makefile:
```diff
@@ -0,0 +1,22 @@
+
+check_port:
+
+	#if ss -tulnp | grep 31346 | grep -q rwslotmachine3; then \
+	if ss -tulnp | grep 31346 | grep -q ld-linux.so.2; then \
+		echo "executable listens or operates on port 31346."; \
+	else \
+		echo "error: executable doesn't operate on port 31346."; \
+ 	fi
```
Why is this commented out?

razvand commented on chapters/mitigations-and-defensive-strategies/defense-mechanisms/activities/03-rwslotmachine3/sol/Makefile:
```diff
@@ -0,0 +1,22 @@
+	gcc $(CFLAGS) rwslotmachine3.c -o rwslotmachine3
+
+check_binary:
+
```
```suggestion
```

razvand commented on chapters/mitigations-and-defensive-strategies/defense-mechanisms/activities/03-rwslotmachine3/sol/Makefile:
```diff
@@ -0,0 +1,22 @@
+CFLAGS = -m32 -fno-stack-protector -Wall -Wno-format-security -no-pie
```
```suggestion
CFLAGS = -m32 -fno-stack-protector -Wall -Wno-format-security
LDFLAGS = -no-pie
```

`-no-pie` is a linker flag.

razvand commented on chapters/mitigations-and-defensive-strategies/defense-mechanisms/activities/03-rwslotmachine3/deploy/Dockerfile:
```diff
@@ -0,0 +1,56 @@
+
+COPY src/ld-linux.so.2 /home/${USER_NAME}/${CHALLENGE_NAME}/ld-linux.so.2
+COPY src/libc.so.6 /home/${USER_NAME}/${CHALLENGE_NAME}/libc.so.6
+
+
```
Any reason for a double blank line?

razvand reviewed (changes requested):
Create two properly worded commits. See instructions [here](https://cbea.ms/git-commit/):
- One commit adds `Dockerfile` and corresponding `Makefile`.
- One commit adds `Makefile` for solution.

cosmintanasa47 committed:
03-rwslotmachine: Add Makefile to be uploaded into container

This Makefile is used inside the containers
to build the executable, check if the executable
exists and also to check if the executable
is running on port 31346.

Resolves: #50

Signed-off-by: Cosmin Tanasa <tanasa.cosminandrei@gmail.com>

razvand reviewed (changes requested):
Use present tense.

Use your full name in the commit messages, i.e. use `Cosmin Tanasa <tanasa.cosminandrei@gmail.com>` instead of `cosmintanasa47 <tanasa.cosminandrei@gmail.com>`

razvand reviewed (changes requested):
Add `README.md` detailing how to run / deploy the container. Nothing out of the ordinary. See sample README from [here](https://github.com/open-education-hub/binary-security/pull/55/files).

cosmintanasa47 committed:
03-rwslotmachine: Add README.md file

README file so everyone knows how to use the containers.

Resolves: #50

Signed-off-by: Cosmin Tanasa <tanasa.cosminandrei@gmail.com>

razvand commented:
@cosmintanasa47 , use athe `03-rwslotmachine:` prefix for your commit messages. Make them look consistent. If you add `Resolves #50`, add it everywhere.

razvand reviewed (changes requested):
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
